### PR TITLE
Collect transaction and contract module/instance metrics

### PR DESCRIPTION
**Pull request by soerenbf**
## Purpose

Adding onto collecting account metrics, this PR adds

- account transactions
- contract modules deployed
- contracts instantiated

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.

### CODE REVIEW

File: kpi-tracker/src/main.rs [hunk 1]
```diff
@@ -1,18 +1,27 @@
 use std::collections::{BTreeMap, HashMap};
 
-use anyhow::Context;
+use anyhow::{anyhow, Context};
 use chrono::{DateTime, Utc};
 use clap::Parser;
+use concordium_rust_sdk::smart_contracts::common::Amount;
+use concordium_rust_sdk::types::hashes::TransactionHash;
+use concordium_rust_sdk::types::smart_contracts::ModuleRef;
+use concordium_rust_sdk::types::{
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:9]: Please use a formatter using the settings we have. This will join these lines.

File: kpi-tracker/src/main.rs [hunk 2]
```diff
@@ -50,8 +50,47 @@ struct BlockDetails {
...
+    is_initial: bool,
+    /// FK to the block in which the account was created.
+    block_hash: BlockHash,
+}
+
+/// Holds selected attributes of an account transaction.
+#[derive(Debug)]
+struct TransactionDetails {
+    /// The transaction type of the account transaction
+    transaction_type: TransactionType,
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:66]: What do you do with the "unknown" one?

File: kpi-tracker/src/main.rs [hunk 3]
```diff
@@ -50,8 +50,47 @@ struct BlockDetails {
...
+    /// FK to the block in which the transaction was finalized.
+    block_hash: BlockHash,
+    /// The cost of the transaction.
+    cost: Amount,
+}
+
+/// Holds selected attributes of a contract module deployed on chain.
+#[derive(Debug)]
+struct ContractModuleDetails {
+    /// FK to the block in which the module was deployed.
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:76]: What is "FK"? ForeignKey? Please do not use abbreviations like this that are unclear.

File: kpi-tracker/src/main.rs [hunk 4]
```diff
@@ -134,69 +187,143 @@ async fn accounts_in_block(
...
+        })
+}
+
+/// Maps `BlockItemSummary` to `BlockEvent`, which represent entities stored in the database.
+fn to_block_events(block_hash: BlockHash, block_item: BlockItemSummary) -> Vec<BlockEvent> {
+    let mut events: Vec<BlockEvent> = Vec::new();
+
+    match block_item.details {
+        AccountTransaction(atd) => {
+            // TODO: do we want to store failed transactions?
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:211]: Yes. That would be a good metric on its own :)

File: kpi-tracker/src/main.rs [hunk 5]
```diff
@@ -134,69 +187,143 @@ async fn accounts_in_block(
...
+            if let Some(details) = get_account_transaction_details(&atd, block_hash) {
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:212]: Let's not ignore failed transctions.

Comment (abizjak) [reply kpi-tracker/src/main.rs:212]: Especially not silently :)

File: kpi-tracker/src/main.rs [hunk 6]
```diff
@@ -134,69 +187,143 @@ async fn accounts_in_block(
...
 /// Insert as a single DB transaction to facilitate easy recovery, as the service can restart from
 /// current height stored in DB.
 fn update_db(
     db: &mut DB,
-    (block_hash, block_details): (BlockHash, &BlockDetails),
+    (block_hash, block_details): (BlockHash, BlockDetails),
     accounts: BTreeMap<AccountAddress, AccountDetails>,
+    transactions: Option<BTreeMap<TransactionHash, TransactionDetails>>,
+    contract_modules: Option<BTreeMap<ModuleRef, ContractModuleDetails>>,
+    contract_instances: Option<BTreeMap<ContractAddress, ContractInstanceDetails>>,
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:301]: I'm not sure what the meaning of all these options is.

Comment (soerenbf) [reply kpi-tracker/src/main.rs:301]: the purpose is to be able to update the database with only accounts (i.e. initialize DB with accounts in genesis block). Maybe it's a bit silly. Could just add them outside of the function in that single case.

File: kpi-tracker/src/main.rs (hunk 2 above)
Comment (soerenbf) [reply kpi-tracker/src/main.rs:66]: Stored this as an option, which I'm thinking will translate to `null` in the database.

Comment (abizjak) [reply kpi-tracker/src/main.rs:66]: Possibly :) You'll have to define the storage type for TransactionType though.

File: kpi-tracker/src/main.rs [hunk 7]
```diff
@@ -22,56 +26,105 @@ struct Args {
...
+struct TransactionDetails {
+    /// The transaction type of the account transaction. Can be none if
+    /// transaction was rejected due to serialization failure.
+    transaction_type: Option<TransactionType>,
+    /// Foreign key to the block in which the transaction was finalized.
+    block_hash:       BlockHash,
+    /// The cost of the transaction.
+    cost:             Amount,
+    /// Whether the transaction failed or not.
+    is_failed:        bool,
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:69]: I would probably have a `is_succes` field instead. I find it weird to focus on the negative.

File: kpi-tracker/src/main.rs [hunk 8]
```diff
@@ -134,179 +187,370 @@ async fn accounts_in_block(
...
-/// Returns a Map of AccountAddress, AccountDetails created in the block represented by `block_hash`
-async fn new_accounts_in_block(
+/// Maps `AccountTransactionDetails` to `TransactionDetails`, where rejected
+/// transactions without a transaction type are represented by `None`.
+fn get_account_transaction_details(
+    details: &AccountTransactionDetails,
+    block_hash: BlockHash,
+) -> TransactionDetails {
+    let transaction_type = details.transaction_type();
+    let is_failed = details.effects.is_rejected().is_some();
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:197]:
```suggestion
    let is_success = details.effects.is_rejected().is_none();
```

File: kpi-tracker/src/main.rs [hunk 9]
```diff
@@ -134,179 +187,370 @@ async fn accounts_in_block(
...
+
+    TransactionDetails {
+        block_hash,
+        transaction_type,
+        is_failed,
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:202]:
```suggestion
        is_success,
```

File: kpi-tracker/src/main.rs [hunk 10]
```diff
@@ -134,179 +187,370 @@ async fn accounts_in_block(
...
+                        module_ref: data.origin_ref,
+                    };
+                    println!(
+                        "CONTRACT INSTANCE:\naddress: {}\ndetails: {:?}",
+                        &data.address, &details
+                    ); // Logger debug
+                    let event = BlockEvent::ContractInstantiation(data.address, details);
+                    events.push(event);
+                }
+                _ => {}
```
Comment (abizjak) [inline kpi-tracker/src/main.rs:244]: I guess "most active smart contracts" come later :)

Comment (soerenbf) [reply kpi-tracker/src/main.rs:244]: #25 😉

**Pull request merged**